File: src/main/java/dev/tuhkanens/comfortlib/implementation/DatabaseImpl.java
package dev.tuhkanens.comfortlib.implementation;

import dev.tuhkanens.comfortlib.ComfortAPI;
import dev.tuhkanens.comfortlib.api.DatabaseAPI;
import dev.tuhkanens.comfortlib.database.base.DatabaseBase;
import dev.tuhkanens.comfortlib.database.base.DatabaseConfig;
import dev.tuhkanens.comfortlib.database.base.DatabaseData;
import dev.tuhkanens.comfortlib.database.base.DatabaseType;
import dev.tuhkanens.comfortlib.database.base.provider.MySQLBase;
import dev.tuhkanens.comfortlib.database.base.provider.SQLiteBase;
import dev.tuhkanens.comfortlib.database.table.Table;
import org.jdbi.v3.core.Jdbi;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

public class DatabaseImpl implements DatabaseAPI {

    private final ConcurrentHashMap<DatabaseType, DatabaseConfig> configs = new ConcurrentHashMap<>();
    private final ConcurrentHashMap<DatabaseType, DatabaseData> bases = new ConcurrentHashMap<>();
    private final List<Table> tables = new ArrayList<>();

    @Override
    public void setDatabases(Collection<DatabaseConfig> configs) {
        setDatabases(false, configs);
    }

    @Override
    public void setForceDatabases(Collection<DatabaseConfig> configs) {
        setDatabases(true, configs);
    }

    private void setDatabases(boolean isForced, Collection<DatabaseConfig> configs) {
        for (DatabaseConfig config : configs) {
            DatabaseType type;
            switch (config) {
                case DatabaseConfig.Sqlite sqlite -> {
                    if (!isForced && !isProvider(DatabaseType.SQLITE)) continue;
                    DatabaseBase base = new SQLiteBase(sqlite.directory());
                    type = DatabaseType.SQLITE;
                    this.bases.put(type, new DatabaseData(base, isForced));
                }
                case DatabaseConfig.Mysql mysql -> {
                    if (!isForced && !isProvider(DatabaseType.MYSQL)) continue;
                    DatabaseBase base = new MySQLBase();
                    type = DatabaseType.MYSQL;
                    this.bases.put(type, new DatabaseData(base, isForced));
                }
            }
            this.configs.put(type, config);
        }
    }

    @Override
    public void setTables(Collection<Table> tables) {
        this.tables.clear();
        this.tables.addAll(tables);
    }

    @Override
    public void setForce(DatabaseType type, boolean isForced) {
        getDataBase(type).setForced(isForced);
    }

    @Override
    public List<Table> getTables() {
        return tables;
    }

    @Override
    public List<DatabaseType> getTypes() {
        return new ArrayList<>(configs.keySet());
    }

    @Override
    public DatabaseConfig getConfig(DatabaseType type) {
        return configs.get(type);
    }

    @Override
    public DatabaseData getDataBase(DatabaseType type) {
        return bases.get(type);
    }

    @Override
    public DatabaseBase getBase(DatabaseType type) {
        return getDataBase(type).getBase();
    }

    @Override
    public Jdbi getDatabase(DatabaseType type) {
        return getBase(type).getDatabase();
    }

    @Override
    public List<DatabaseData> getDataBases() {
        return new ArrayList<>(bases.values());
    }

    @Override
    public List<DatabaseBase> getAllBases() {
        return getDataBases()
                .stream()
                .map(DatabaseData::getBase)
                .toList();
    }

    @Override
    public List<DatabaseBase> getUnforcedBases() {
        return getDataBases()
                .stream()
                .filter(base -> !base.getForced())
                .map(DatabaseData::getBase)
                .toList();
    }

    @Override
    public List<DatabaseBase> getForcedBases() {
        return getDataBases()
                .stream()
                .filter(DatabaseData::getForced)
                .map(DatabaseData::getBase)
                .toList();
    }

    @Override
    public DatabaseType getProvider() {
        String rawProvider = ComfortAPI.CONFIG
                .getNode()
                .node("database", "provider")
                .getString("sqlite")
                .toUpperCase();
        try {
            return DatabaseType.valueOf(rawProvider);
        } catch (IllegalArgumentException e) {
            return DatabaseType.SQLITE;
        }
    }

    @Override
    public boolean isProvider(DatabaseType type) {
        return getProvider() == type;
    }

    @Override
    public boolean getForced(DatabaseType type) {
        return getDataBase(type).getForced();
    }

    @Override
    public void connect() {
        getAllBases().forEach(DatabaseBase::connect);

        if (tables.isEmpty()) return;

        for (DatabaseBase base : getAllBases()) {
            Jdbi database = base.getDatabase();
            for (Table table : tables) {
                table.create(database);
            }
        }
    }

    @Override
    public void disconnect() {
        for (DatabaseType type : DatabaseType.values()) {
            getBase(type).disconnect();
        }
    }

    @Override
    public void execute(Jdbi database, String sql) {
        database.useHandle(handle -> handle.execute(sql));
    }

    @Override
    public void execute(DatabaseType type, String sql) {
        execute(getDatabase(type), sql);
    }

    @Override
    public void execute(Collection<DatabaseType> types, String sql) {
        for (DatabaseType type : types) {
            execute(type, sql);
        }
    }

    @Override
    public void execute(Collection<DatabaseType> types, boolean isForced, String sql) {
        for (DatabaseType type : types) {
            if (getForced(type) == isForced) {
                execute(type, sql);
            }
        }
    }

    @Override
    public void execute(Jdbi database, Collection<String> sql) {
        for (String s : sql) {
            execute(database, s);
        }
    }

    @Override
    public void execute(DatabaseType type, Collection<String> sql) {
        execute(getDatabase(type), sql);
    }

    @Override
    public void execute(Collection<DatabaseType> types, Collection<String> sql) {
        for (DatabaseType type : types) {
            execute(type, sql);
        }
    }

    @Override
    public void execute(Collection<DatabaseType> types, boolean isForced, Collection<String> sql) {
        for (DatabaseType type : types) {
            if (getForced(type) == isForced) {
                execute(type, sql);
            }
        }
    }
}

File: src/main/kotlin/dev/tuhkanens/comfortlib/api/implementation/DatabaseImpl.kt
package dev.tuhkanens.comfortlib.api.implementation

import dev.tuhkanens.comfortlib.ComfortAPI
import dev.tuhkanens.comfortlib.api.ConfigAPI
import dev.tuhkanens.comfortlib.api.DatabaseAPI
import dev.tuhkanens.comfortlib.database.DatabaseBase
import dev.tuhkanens.comfortlib.database.DatabaseConfig
import dev.tuhkanens.comfortlib.database.DatabaseData
import dev.tuhkanens.comfortlib.database.DatabaseType
import dev.tuhkanens.comfortlib.database.base.MySQLBase
import dev.tuhkanens.comfortlib.database.base.SQLiteBase
import org.jetbrains.exposed.v1.core.Table
import org.jetbrains.exposed.v1.jdbc.Database
import org.jetbrains.exposed.v1.jdbc.SchemaUtils
import org.jetbrains.exposed.v1.jdbc.transactions.transaction
import org.jetbrains.exposed.v1.migration.jdbc.MigrationUtils
import java.util.concurrent.ConcurrentHashMap

class DatabaseImpl : DatabaseAPI {

    private val configs: ConcurrentHashMap<DatabaseType, DatabaseConfig> = ConcurrentHashMap()
    private val bases: ConcurrentHashMap<DatabaseType, DatabaseData> = ConcurrentHashMap()
    private val tables: MutableSet<Table> = ConcurrentHashMap.newKeySet()

    override fun setDatabases(vararg config: DatabaseConfig) {
        setDatabases(*config, force = false)
    }

    override fun setForceDatabases(vararg config: DatabaseConfig) {
        setDatabases(*config, force = true)
    }

    private fun setDatabases(vararg config: DatabaseConfig, force: Boolean = false) {
        if (config.isEmpty()) return
        for (c in config) {
            val type = when (c) {
                is DatabaseConfig.Sqlite -> {
                    if (!force && !isProvider(DatabaseType.SQLITE)) return
                    val base = SQLiteBase(c.directory)
                    bases[DatabaseType.SQLITE] = DatabaseData(base, force)
                    DatabaseType.SQLITE
                }
                is DatabaseConfig.Mysql -> {
                    if (!force && !isProvider(DatabaseType.MYSQL)) return
                    val base = MySQLBase()
                    bases[DatabaseType.MYSQL] = DatabaseData(base, force)
                    DatabaseType.MYSQL
                }
            }
            this.configs[type] = c
        }
    }

    override fun setTables(vararg table: Table) {
        this.tables.addAll(table)
    }

    override fun setForce(type: DatabaseType, isForced: Boolean) {
        getBaseData(type)?.isForced = isForced
    }

    override fun getTables(): Array<Table> {
        return tables.toTypedArray()
    }

    override fun getTypes(): List<DatabaseType> {
        return bases.keys.toList()
    }

    override fun getConfig(type: DatabaseType): DatabaseConfig? {
        return configs[type]
    }

    override fun getAllBases(): List<DatabaseBase> {
        return bases.values
            .map { it.base }
    }

    override fun getUnforcedBases(): List<DatabaseBase> {
        return bases.values
            .filter { !it.isForced }
            .map { it.base }
    }

    override fun getForcedBases(): List<DatabaseBase> {
        return bases.values
            .filter { it.isForced }
            .map { it.base }
    }

    override fun getBaseData(type: DatabaseType): DatabaseData? {
        return bases[type]
    }

    override fun getBase(type: DatabaseType): DatabaseBase? {
        return getBaseData(type)?.base
    }

    override fun getDatabase(type: DatabaseType): Database? {
        return getBase(type)?.getDatabase()
    }

    override fun getProvider(): DatabaseType {
        val rawProvider = ComfortAPI.get<ConfigAPI>()
            .getNode()
            .node("database", "provider")
            .getString("sqlite")
            .uppercase()
        return runCatching { DatabaseType.valueOf(rawProvider) }
            .getOrDefault(DatabaseType.SQLITE)
    }

    override fun isProvider(type: DatabaseType): Boolean {
        return getProvider() == type
    }

    override fun isForced(type: DatabaseType): Boolean {
        return getBaseData(type)?.isForced ?: false
    }

    override fun connect() {
        bases.values.forEach { it.base.connect() }

        if (tables.isNotEmpty()) {
            val tablesArray = tables.toTypedArray()
            for (data in bases.values) {
                val db = data.base.getDatabase()
                transaction(db) {
                    val statements = MigrationUtils.statementsRequiredForDatabaseMigration(*tablesArray)
                    if (statements.isNotEmpty()) {
                        statements.forEach { exec(it) }
                    } else {
                        SchemaUtils.create(*tablesArray)
                    }
                }
            }
        }
    }

    override fun disconnect() {
        for (type in DatabaseType.entries) {
            getBase(type)?.disconnect()
        }
    }

    override fun <T> transaction(type: DatabaseType, block: () -> T): T {
        val db = getDatabase(type) ?: throw IllegalStateException("Database $type is not connected")
        return transaction(db) {
            block()
        }
    }

    override fun <T> transaction(vararg type: DatabaseType, force: Boolean, block: () -> T): T {
        if (type.isEmpty()) {
            return block()
        }
        fun executeNested(index: Int): T {
            if (index >= type.size) {
                return block()
            }
            val currentType = type[index]
            val db = getDatabase(currentType) ?: throw IllegalStateException("Database $currentType is not connected")
            val data = getBaseData(currentType) ?: throw IllegalStateException("DatabaseData for $currentType not found")

            if (force != data.isForced) {
                return executeNested(index + 1)
            }

            return transaction(db) {
                executeNested(index + 1)
            }
        }

        return executeNested(0)
    }

    override fun <T> transaction(force: Boolean, block: () -> T): T {
        val bases = bases.keys.toTypedArray()
        return transaction(*bases, force = force) {
            block()
        }
    }

}